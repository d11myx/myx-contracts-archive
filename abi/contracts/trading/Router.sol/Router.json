[
  {
    "inputs": [
      {
        "internalType": "contract IAddressesProvider",
        "name": "_addressProvider",
        "type": "address"
      },
      {
        "internalType": "contract ITradingRouter",
        "name": "_tradingRouter",
        "type": "address"
      },
      {
        "internalType": "contract IOrderManager",
        "name": "_orderManager",
        "type": "address"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "constructor"
  },
  {
    "anonymous": false,
    "inputs": [
      {
        "indexed": false,
        "internalType": "address",
        "name": "oldAddress",
        "type": "address"
      },
      {
        "indexed": false,
        "internalType": "address",
        "name": "newAddress",
        "type": "address"
      }
    ],
    "name": "UpdateTradingRouter",
    "type": "event"
  },
  {
    "inputs": [],
    "name": "addressProvider",
    "outputs": [
      {
        "internalType": "contract IAddressesProvider",
        "name": "",
        "type": "address"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "pairIndex",
        "type": "uint256"
      },
      {
        "internalType": "bool",
        "name": "isLong",
        "type": "bool"
      }
    ],
    "name": "cancelAllPositionOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "cancelDecreaseOrder",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "cancelIncreaseOrder",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "pairIndex",
        "type": "uint256"
      },
      {
        "internalType": "bool",
        "name": "isLong",
        "type": "bool"
      },
      {
        "internalType": "bool",
        "name": "isIncrease",
        "type": "bool"
      }
    ],
    "name": "cancelOrders",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "triggerPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          }
        ],
        "internalType": "struct TradingTypes.DecreasePositionRequest",
        "name": "_request",
        "type": "tuple"
      }
    ],
    "name": "createDecreaseOrder",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.IncreasePositionRequest",
        "name": "_request",
        "type": "tuple"
      }
    ],
    "name": "createIncreaseOrder",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.CreateTpSlRequest",
        "name": "request",
        "type": "tuple"
      }
    ],
    "name": "createTpSl",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "tpOrderId",
        "type": "uint256"
      },
      {
        "internalType": "uint256",
        "name": "slOrderId",
        "type": "uint256"
      }
    ],
    "stateMutability": "nonpayable",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "index",
        "type": "uint256"
      }
    ],
    "name": "decreaseLimitOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "triggerPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "abovePrice",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "needADL",
            "type": "bool"
          }
        ],
        "internalType": "struct TradingTypes.DecreasePositionOrder",
        "name": "",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "decreaseLimitOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "decreaseMarketOrderStartIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "index",
        "type": "uint256"
      }
    ],
    "name": "decreaseMarketOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "triggerPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "abovePrice",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "needADL",
            "type": "bool"
          }
        ],
        "internalType": "struct TradingTypes.DecreasePositionOrder",
        "name": "",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "decreaseMarketOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "_orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "_tradeType",
        "type": "uint8"
      }
    ],
    "name": "getDecreaseOrder",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "triggerPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "abovePrice",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "needADL",
            "type": "bool"
          }
        ],
        "internalType": "struct TradingTypes.DecreasePositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "_orderId",
        "type": "uint256"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "_tradeType",
        "type": "uint8"
      }
    ],
    "name": "getIncreaseOrder",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.IncreasePositionOrder",
        "name": "order",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "key",
        "type": "bytes32"
      }
    ],
    "name": "getPositionOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "bool",
            "name": "isIncrease",
            "type": "bool"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.PositionOrder[]",
        "name": "orders",
        "type": "tuple[]"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "index",
        "type": "uint256"
      }
    ],
    "name": "increaseLimitOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.IncreasePositionOrder",
        "name": "",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "increaseLimitOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "increaseMarketOrderStartIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "uint256",
        "name": "index",
        "type": "uint256"
      }
    ],
    "name": "increaseMarketOrders",
    "outputs": [
      {
        "components": [
          {
            "internalType": "uint256",
            "name": "orderId",
            "type": "uint256"
          },
          {
            "internalType": "address",
            "name": "account",
            "type": "address"
          },
          {
            "internalType": "uint256",
            "name": "pairIndex",
            "type": "uint256"
          },
          {
            "internalType": "enum TradingTypes.TradeType",
            "name": "tradeType",
            "type": "uint8"
          },
          {
            "internalType": "int256",
            "name": "collateral",
            "type": "int256"
          },
          {
            "internalType": "uint256",
            "name": "openPrice",
            "type": "uint256"
          },
          {
            "internalType": "bool",
            "name": "isLong",
            "type": "bool"
          },
          {
            "internalType": "uint256",
            "name": "sizeAmount",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tpPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "tp",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "slPrice",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "sl",
            "type": "uint256"
          },
          {
            "internalType": "uint256",
            "name": "blockTime",
            "type": "uint256"
          }
        ],
        "internalType": "struct TradingTypes.IncreasePositionOrder",
        "name": "",
        "type": "tuple"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "increaseMarketOrdersIndex",
    "outputs": [
      {
        "internalType": "uint256",
        "name": "",
        "type": "uint256"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "orderManager",
    "outputs": [
      {
        "internalType": "contract IOrderManager",
        "name": "",
        "type": "address"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "bytes32",
        "name": "positionKey",
        "type": "bytes32"
      },
      {
        "internalType": "enum TradingTypes.TradeType",
        "name": "tradeType",
        "type": "uint8"
      }
    ],
    "name": "positionHasTpSl",
    "outputs": [
      {
        "internalType": "bool",
        "name": "",
        "type": "bool"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [],
    "name": "tradingRouter",
    "outputs": [
      {
        "internalType": "contract ITradingRouter",
        "name": "",
        "type": "address"
      }
    ],
    "stateMutability": "view",
    "type": "function"
  },
  {
    "inputs": [
      {
        "internalType": "contract ITradingRouter",
        "name": "_tradingRouter",
        "type": "address"
      }
    ],
    "name": "updateTradingRouter",
    "outputs": [],
    "stateMutability": "nonpayable",
    "type": "function"
  }
]
