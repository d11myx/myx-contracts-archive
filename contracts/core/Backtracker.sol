// SPDX-License-Identifier: BUSL-1.1
pragma solidity 0.8.19;

import "../interfaces/IBacktracker.sol";

contract Backtracker is IBacktracker {

    event Backtracking(address account, uint64 round);

    event UnBacktracking(address account);

    bool public override backtracking;
    uint64 public override backtrackRound;

    constructor() {
        backtracking = false;
    }

    modifier whenNotBacktracking() {
        _requireNotBacktracking();
        _;
    }

    modifier whenBacktracking() {
        _requireBacktracking();
        _;
    }

    function enterBacktracking(uint64 _backtrackRound) external whenNotBacktracking {
        backtracking = true;
        backtrackRound = _backtrackRound;
        emit Backtracking(msg.sender, _backtrackRound);
    }

    function quitBacktracking() external whenBacktracking {
        backtracking = false;
        emit UnBacktracking(msg.sender);
    }

    function _requireNotBacktracking() internal view {
        require(!backtracking, "Backtracker: backtracking");
    }

    function _requireBacktracking() internal view {
        require(backtracking, "Backtracker: not backtracking");
    }
}
