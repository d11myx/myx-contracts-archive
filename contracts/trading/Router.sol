// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.17;

import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";

import "../interfaces/IRouter.sol";
import "../interfaces/IAddressesProvider.sol";
import "../interfaces/IRoleManager.sol";
import "./interfaces/ITradingRouter.sol";
import "hardhat/console.sol";

contract Router is IRouter, ReentrancyGuardUpgradeable {

    IAddressesProvider public immutable addressProvider;

    ITradingRouter public tradingRouter;

    modifier onlyPoolAdmin() {
        require(IRoleManager(addressProvider.getRoleManager()).isPoolAdmin(msg.sender), "onlyPoolAdmin");
        _;
    }

    constructor(IAddressesProvider _addressProvider, ITradingRouter _tradingRouter) {
        addressProvider = _addressProvider;
        tradingRouter = _tradingRouter;
    }

    function updateTradingRouter(ITradingRouter _tradingRouter) external override onlyPoolAdmin {
        address oldAddress = address(_tradingRouter);
        tradingRouter = _tradingRouter;
        address newAddress = address(tradingRouter);

        emit UpdateTradingRouter(oldAddress, newAddress);
    }

    function increaseMarketOrdersIndex() external view override returns (uint256) {
        return tradingRouter.increaseMarketOrdersIndex();
    }

    function decreaseMarketOrdersIndex() external view override returns (uint256) {
        return tradingRouter.decreaseMarketOrdersIndex();
    }

    function increaseMarketOrderStartIndex() external view override returns (uint256) {
        return tradingRouter.increaseMarketOrderStartIndex();
    }

    function decreaseMarketOrderStartIndex() external view override returns (uint256) {
        return tradingRouter.decreaseMarketOrderStartIndex();
    }

    function increaseLimitOrdersIndex() external view override returns (uint256) {
        return tradingRouter.increaseLimitOrdersIndex();
    }

    function decreaseLimitOrdersIndex() external view override returns (uint256) {
        return tradingRouter.decreaseLimitOrdersIndex();
    }

    function positionHasTpSl(bytes32 positionKey, ITradingRouter.TradeType tradeType) external view override returns (bool) {
        return tradingRouter.positionHasTpSl(positionKey, tradeType);
    }

    function createIncreaseOrder(ITradingRouter.IncreasePositionRequest memory _request) external override nonReentrant returns (uint256) {
        return tradingRouter.createIncreaseOrder(_request);
    }

    function cancelIncreaseOrder(uint256 _orderId, ITradingRouter.TradeType _tradeType) external override nonReentrant {
        tradingRouter.cancelIncreaseOrder(_orderId, _tradeType);
    }

    function createDecreaseOrder(ITradingRouter.DecreasePositionRequest memory _request) external override nonReentrant returns (uint256) {
        return tradingRouter.createDecreaseOrder(_request);
    }

    function cancelDecreaseOrder(uint256 _orderId, ITradingRouter.TradeType _tradeType) external override nonReentrant {
        tradingRouter.cancelDecreaseOrder(_orderId, _tradeType);
    }

    function cancelAllPositionOrders(address account, uint256 pairIndex, bool isLong) external override {
        tradingRouter.cancelAllPositionOrders(account, pairIndex, isLong);
    }

    function cancelOrders(address account, uint256 pairIndex, bool isLong, bool isIncrease) external override {
        tradingRouter.cancelOrders(account, pairIndex, isLong, isIncrease);
    }

    function createTpSl(ITradingRouter.CreateTpSlRequest memory _request) external override returns (uint256 tpOrderId, uint256 slOrderId) {
        return tradingRouter.createTpSl(_request);
    }

    function getIncreaseOrder(uint256 _orderId, ITradingRouter.TradeType _tradeType) external override view returns (ITradingRouter.IncreasePositionOrder memory order) {
        return tradingRouter.getIncreaseOrder(_orderId, _tradeType);
    }

    function getDecreaseOrder(uint256 _orderId, ITradingRouter.TradeType _tradeType) external override view returns (ITradingRouter.DecreasePositionOrder memory order) {
        return tradingRouter.getDecreaseOrder(_orderId, _tradeType);
    }

    function getPositionOrders(bytes32 key) external override view returns (ITradingRouter.PositionOrder[] memory orders) {
        return tradingRouter.getPositionOrders(key);
    }

}
