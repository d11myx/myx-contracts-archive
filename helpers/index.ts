export * from './constants';
export * from './contract-deployments';
export * from './contract-getters';
export * from './deploy-ids';
export * from './env';
export * from './hardhat-config-helpers';
export * from './init-helper';
export * from './market-config-helper';
export * from './types';
export * from './utilities/signer';
export * from './utilities/tx';
export * from '../tasks/misc/print-deployments';

import { loadTasks } from './hardhat-config-helpers';

const TASK_FOLDERS = ['../tasks/misc'];

// Load all plugin tasks
loadTasks(TASK_FOLDERS);
