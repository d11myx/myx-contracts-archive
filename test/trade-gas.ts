import { TestEnv, newTestEnv } from './helpers/make-suite';
import { ethers } from 'hardhat';
import { expect } from './shared/expect';
import { deployMockCallback, MAX_UINT_AMOUNT, TradeType, waitForTx } from '../helpers';
import { mintAndApprove, updateBTCPrice } from './helpers/misc';
import snapshotGasCost from './shared/snapshotGasCost';
import { BigNumber } from 'ethers';
import { TradingTypes } from '../types/contracts/core/Router';

describe('Router: increase position ar', () => {
    const pairIndex = 0;
    let localTestEnv: TestEnv;
    let orderId: BigNumber;

    before(async () => {
        localTestEnv = (await newTestEnv()) as TestEnv;
        const {
            deployer,
            users: [depositor, poolAdmin, operator],
            btc,
            usdt,
            pool,
            roleManager,
            oraclePriceFeed,
        } = localTestEnv;

        const pair = await pool.getPair(pairIndex);

        await roleManager.connect(deployer.signer).addOperator(operator.address);
        await roleManager.connect(operator.signer).removeAccountBlackList(depositor.address);

        await updateBTCPrice(localTestEnv, '30000');
    });

    it('addLiquidity cast', async () => {
        const {
            deployer,
            users: [depositor, poolAdmin, operator, trader],
            keeper,
            router,
            btc,
            usdt,
            pool,
            positionManager,
            executor,
            orderManager,
        } = localTestEnv;
        let testCallBack = await deployMockCallback();
        const pair = await pool.getPair(pairIndex);
        const indexAmount = ethers.utils.parseUnits('10000', 18);
        const stableAmount = ethers.utils.parseUnits('300000000', 18);
        await mintAndApprove(localTestEnv, btc, indexAmount, depositor, testCallBack.address);
        await mintAndApprove(localTestEnv, usdt, stableAmount, depositor, testCallBack.address);

        await testCallBack
            .connect(depositor.signer)
            .addLiquidity(pool.address, pair.indexToken, pair.stableToken, indexAmount, stableAmount);
    });

    it('createIncreaseOrderWithoutTpSl cast', async () => {
        const {
            deployer,
            users: [depositor, poolAdmin, operator, trader],
            keeper,
            router,
            btc,
            usdt,
            pool,
            positionManager,
            executor,
            orderManager,
        } = localTestEnv;
        const amount = ethers.utils.parseUnits('30000', 18);
        await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, amount));

        // View user's position
        let traderPosition = await positionManager.getPosition(trader.address, pairIndex, true);
        // console.log("user's position", traderPosition);

        const collateral = ethers.utils.parseUnits('10000', 18);
        await waitForTx(await usdt.connect(deployer.signer).mint(trader.address, collateral));
        await usdt.connect(trader.signer).approve(router.address, MAX_UINT_AMOUNT);

        const increasePositionRequest: TradingTypes.IncreasePositionRequestStruct = {
            account: trader.address,
            pairIndex: pairIndex,
            tradeType: TradeType.MARKET,
            collateral: collateral,
            openPrice: ethers.utils.parseUnits('30000', 30),
            isLong: true,
            sizeAmount: ethers.utils.parseUnits('5', 18),
            maxSlippage: 0,
        };

        orderId = await orderManager.ordersIndex();

        // console.log(`order:`, await orderManager.increaseMarketOrders(orderId));

        await snapshotGasCost(router.connect(trader.signer).createIncreaseOrderWithoutTpSl(increasePositionRequest));
    });
    it('executeIncreaseOrder cast', async () => {
        const {
            deployer,
            users: [depositor, poolAdmin, operator, trader],
            keeper,
            router,
            btc,
            usdt,
            pool,
            positionManager,
            executionLogic,
            orderManager,
        } = localTestEnv;
        await snapshotGasCost(
            executionLogic.connect(keeper.signer).executeIncreaseOrder(orderId, TradeType.MARKET, 0, 0),
        );
    });
    it('createDecreaseOrder cast', async () => {
        const {
            deployer,
            users: [depositor, poolAdmin, operator, trader],
            keeper,
            router,
            btc,
            usdt,
            pool,
            positionManager,
            executionLogic,
            orderManager,
        } = localTestEnv;
        const position = await positionManager.getPosition(trader.address, pairIndex, true);
        // console.log(`position:`, position);

        expect(position.positionAmount).to.be.eq(ethers.utils.parseUnits('5', 18));

        const decreasePositionRequest: TradingTypes.DecreasePositionRequestStruct = {
            account: trader.address,
            pairIndex: pairIndex,
            tradeType: TradeType.MARKET,
            collateral: 0,
            triggerPrice: ethers.utils.parseUnits('30000', 30),
            isLong: true,
            sizeAmount: ethers.utils.parseUnits('5', 18),
            maxSlippage: 0,
        };
        orderId = await orderManager.ordersIndex();
        await snapshotGasCost(router.connect(trader.signer).createDecreaseOrder(decreasePositionRequest));
    });
    it('executeDecreaseOrder cast', async () => {
        const {
            deployer,
            users: [depositor, poolAdmin, operator, trader],
            keeper,
            router,
            btc,
            usdt,
            pool,
            positionManager,
            executionLogic,
            orderManager,
        } = localTestEnv;
        await snapshotGasCost(
            executionLogic.connect(keeper.signer).executeDecreaseOrder(orderId, TradeType.MARKET, 0, 0, false),
        );

        let traderPosition = await positionManager.getPosition(trader.address, pairIndex, true);
        const lastTimePrice = traderPosition.averagePrice;
        // console.log(`before closing position: `, traderPosition);
        // console.log(`price before closing position: `, lastTimePrice);

        const closingPosition = await positionManager.getPosition(trader.address, pairIndex, true);
        const closingPositionPrice = closingPosition.averagePrice;
        // console.log(`afer closing position: `, closingPosition);
        // console.log(`price afer closing position: `, closingPositionPrice);
    });
});
