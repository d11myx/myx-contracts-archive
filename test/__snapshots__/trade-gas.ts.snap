// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router: increase position ar createDecreaseOrder cast 1`] = `425580`;

exports[`Router: increase position ar createIncreaseOrderWithoutTpSl cast 1`] = `448238`;

exports[`Router: increase position ar executeDecreaseOrder cast 1`] = `367545`;

exports[`Router: increase position ar executeIncreaseOrder cast 1`] = `567394`;
